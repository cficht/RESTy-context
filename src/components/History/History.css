@import url('https://fonts.googleapis.com/css?family=Oswald&display=swap');

.History {
  display: flex;
  flex-direction: column;
  align-items: center;
  border: 2px solid #e15e07;
  border-radius: 10px;
  margin: 10px;
  background-color: #ff7315;

  h1 {
    margin-block-start: .5em;
    margin-block-end: .2em;
    font-family: 'Oswald', sans-serif;
    font-size: 2.5em;
  }

  ul {
    list-style-type: none;
    margin-block-start: 0em;
    margin-block-end: 0em;
    margin-inline-start: 0px;
    margin-inline-end: 0px;
    padding-inline-start: 0px;
    width: 80%;
    font-family: 'Oswald', sans-serif;

    h3 {
      margin-block-start: 0em;
      margin-block-end: 0em;
    }

    p {
      margin-block-start: 0em;
      margin-block-end: 0em;
      font-style: italic;
    }

    li {
      padding: 10px;
      margin: 10px 0px 10px 0px;
      border: 2px solid #232020;
      border-radius: 10px;
      background-color: #3a3535;
      color: white;
    }
  
    li:hover {
      background-color: #f48c47;
    }
  }

  button {
    margin: 10px;
    border: 2px solid #232020;
    border-radius: 10px;
    background-color: #3a3535;
    color: white;
  }

  button:hover {
    background-color: #f48c47;
  }
}
