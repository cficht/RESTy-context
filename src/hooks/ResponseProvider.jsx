import React, { createContext, useReducer, useContext } from 'react';
import PropTypes from 'prop-types';

const ResponseContext = createContext();

const initialState = {
  url: '',
  method: '',
  body: '',
  res: {},
  requests: []
};

export function reducer(state, action) {
  switch(action.type) {
    case 'SET_URL':
      return { ...state, url: action.payload };
    case 'SET_METHOD':
      return { ...state, method: action.payload };
    case 'SET_BODY':
      return { ...state, body: action.payload };
    case 'SET_RES':
      return { ...state, res: action.payload };
    case 'ADD_REQUESTS':
      return { ...state, requests: [...state.requests, action.payload] };
    default:
      return state;
  }
}

export const ResponseProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <ResponseContext.Provider value={{ state, dispatch }}>
      {children}
    </ResponseContext.Provider>
  );
};

ResponseProvider.propTypes = {
  children: PropTypes.node
};

export const useGlobalState = () => {
  const { state } = useContext(ResponseContext);
  return state;
};

export const useDispatch = () => {
  const { dispatch } = useContext(ResponseContext);
  return dispatch;
};

export const useUrl = () => {
  const { url } = useGlobalState();
  return url;
};

export const useMethod = () => {
  const { method } = useGlobalState();
  return method;
};

export const useBody = () => {
  const { body } = useGlobalState();
  return body;
};

export const useRes = () => {
  const { res } = useGlobalState();
  return res;
};

export const useRequests = () => {
  const { requests } = useGlobalState();
  return requests;
};
