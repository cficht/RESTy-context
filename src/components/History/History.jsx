import React from 'react';
import { useRequests } from '../../hooks/ResponseProvider';

const History = () => {
  const requests = useRequests();
  const requestNodes = requests.map(request => {
    const home = request.url.split('/')[2];
    const user = request.url.split('.com')[1];
    return (<li key={`${request.url}-${request.method}-${request.body}`}>
      <h3>{request.method}</h3>
      <p>{home}</p>
      <p>{user}</p>
    </li>
    );
  });

  return (
    <>
      <h1>HISTORY</h1>
      <ul>
        {requestNodes}
      </ul>
    </>
  );
};

export default History;
