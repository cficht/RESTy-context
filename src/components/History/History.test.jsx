import React from 'react';
import { shallow } from 'enzyme';
import History from './History';
import { ResponseProvider } from '../../hooks/ResponseProvider';

describe('History component', () => {
  it('renders History', () => {
    const wrapper = shallow(
      <ResponseProvider>
        <History />
      </ResponseProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
