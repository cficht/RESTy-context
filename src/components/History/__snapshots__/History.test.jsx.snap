// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History component renders History 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "handleChange": [Function],
      "handleClear": [Function],
      "handleClick": [Function],
      "handleSubmit": [Function],
      "state": Object {
        "auth": "none",
        "body": "",
        "method": "",
        "password": "",
        "requests": Array [],
        "res": Object {},
        "token": "",
        "url": "",
        "username": "",
      },
    }
  }
>
  <History />
</ContextProvider>
`;
